Fix for PPO2 When loading a model and then training with vectorized environment with a different vector length

**Describe the bug**
Lets say you train PPO2 with a vectorized environment of length 2, then you save and load the model and train with a vectorized environment of size 1. A crash occurs.

**Code example**

```python

    env = make_vec_env('CartPole-v1',n_envs=2)
    model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
    
    model.learn(total_timesteps = 100)
    model.save("ppo2_cartpole")
    
    del model
    
    model = PPO2.load("ppo2_cartpole")
    test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
    
    model.set_env(test_env)
    model.learn(total_timesteps = 100)

```
**Error**
```bash
slices = (arr[mbinds] for arr in (obs, returns, masks, actions, values, neglogpacs))
E   IndexError: index 13 is out of bounds for axis 0 with size 10

stable_baselines/ppo2/ppo2.py:362: IndexError


```

**System Info**
*development build from source(this error should be easy to reproduce with any build)*
*python3.7*

**Additional context**
I have prepped a pull request with the patch and corresponding test. The test will fail when the patch is not applied. It may come down to opinion whether this is an error or not, but I think it is, since n_batch depends on n_envs and there does not seems to be a mechanism for updating n_batch outside of initialization. I am interested to hear the thoughts of the maintainers on this proposed patch.

--- a/docs/misc/changelog.rst
+++ b/docs/misc/changelog.rst
@@ -18,6 +18,7 @@ New Features:
 Bug Fixes:
 ^^^^^^^^^^
 - Fixed bug in pretraining method that prevented from calling it twice.
+- Fixed a bug where a crash would occur if a PPO2 model was trained in a vectorized environment, saved and subsequently loaded, then trained in a vectorized environment with a different length
 
 Deprecations:
 ^^^^^^^^^^^^^

--- a/stable_baselines/a2c/a2c.py
+++ b/stable_baselines/a2c/a2c.py
@@ -241,6 +241,10 @@ class A2C(ActorCriticRLModel):
 
         return policy_loss, value_loss, policy_entropy
 
+    def set_env(self,env):
+        super().set_env(env)
+        self.n_batch = self.n_envs * self.n_steps
+
     def learn(self, total_timesteps, callback=None, log_interval=100, tb_log_name="A2C",
               reset_num_timesteps=True):
 

--- a/stable_baselines/common/callbacks.py
+++ b/stable_baselines/common/callbacks.py
@@ -251,8 +251,8 @@ class EvalCallback(EventCallback):
                  callback_on_new_best: Optional[BaseCallback] = None,
                  n_eval_episodes: int = 5,
                  eval_freq: int = 10000,
-                 log_path: str = None,
-                 best_model_save_path: str = None,
+                 log_path: Optional[str] = None,
+                 best_model_save_path: Optional[str] = None,
                  deterministic: bool = True,
                  render: bool = False,
                  verbose: int = 1):

--- a/stable_baselines/ppo2/ppo2.py
+++ b/stable_baselines/ppo2/ppo2.py
@@ -298,7 +298,11 @@ class PPO2(ActorCriticRLModel):
                 [self.pg_loss, self.vf_loss, self.entropy, self.approxkl, self.clipfrac, self._train], td_map)
 
         return policy_loss, value_loss, policy_entropy, approxkl, clipfrac
-
+    
+    def set_env(self,env):
+        super().set_env(env)
+        self.n_batch = self.n_envs * self.n_steps
+    
     def learn(self, total_timesteps, callback=None, log_interval=1, tb_log_name="PPO2",
               reset_num_timesteps=True):
         # Transform to callable if needed

--- a/tests/test_a2c.py
+++ b/tests/test_a2c.py
@@ -1,7 +1,7 @@
 import os
 
 import pytest
-import gym 
+import gym
 
 from stable_baselines import A2C
 from stable_baselines.common import make_vec_env
@@ -9,17 +9,18 @@ from stable_baselines.common.vec_env import DummyVecEnv
 
 
 def test_a2c_update_n_batch_on_load(tmp_path):
-    env = make_vec_env('CartPole-v1',n_envs=2)
-    model = A2C('MlpPolicy', env, n_steps = 10)
-    
-    model.learn(total_timesteps = 100)
-    model.save(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))
-    
+    env = make_vec_env("CartPole-v1", n_envs=2)
+    model = A2C("MlpPolicy", env, n_steps=10)
+
+    model.learn(total_timesteps=100)
+    model.save(os.path.join(str(tmp_path), "a2c_cartpole.zip"))
+
     del model
-    
-    model = A2C.load(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))
-    test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
-    
+
+    model = A2C.load(os.path.join(str(tmp_path), "a2c_cartpole.zip"))
+    test_env = DummyVecEnv([lambda: gym.make("CartPole-v1")])
+
     model.set_env(test_env)
-    assert(model.n_batch == 10)
-    os.remove(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))
+    assert model.n_batch == 10
+    model.learn(100)
+    os.remove(os.path.join(str(tmp_path), "a2c_cartpole.zip"))

--- a/tests/test_a2c.py
+++ b/tests/test_a2c.py
@@ -1,0 +1,27 @@
+import os
+import pathlib
+
+import pytest
+import gym 
+
+from stable_baselines import A2C
+from stable_baselines.common import make_vec_env
+from stable_baselines.common.vec_env import DummyVecEnv
+
+
+def test_a2c_update_n_batch_on_load(tmp_path):
+    env = make_vec_env('CartPole-v1',n_envs=2)
+    model = A2C('MlpPolicy', env, n_steps = 10)
+    
+    model.learn(total_timesteps = 100)
+    model.save(tmp_path / "a2c_cartpole.zip")
+    
+    del model
+    
+    model = A2C.load(tmp_path / "a2c_cartpole.zip")
+    test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
+    
+    model.set_env(test_env)
+    model.learn(total_timesteps = 100)
+    assert(model.n_batch == 10)
+    os.remove(tmp_path / "a2c_cartpole.zip")

--- a/tests/test_a2c.py
+++ b/tests/test_a2c.py
@@ -1,5 +1,4 @@
 import os
-import pathlib
 
 import pytest
 import gym 
@@ -14,14 +13,13 @@ def test_a2c_update_n_batch_on_load(tmp_path):
     model = A2C('MlpPolicy', env, n_steps = 10)
     
     model.learn(total_timesteps = 100)
-    model.save(tmp_path / "a2c_cartpole.zip")
+    model.save(os.path.join(tmp_path , "a2c_cartpole.zip"))
     
     del model
     
-    model = A2C.load(tmp_path / "a2c_cartpole.zip")
+    model = A2C.load(os.path.join(tmp_path , "a2c_cartpole.zip"))
     test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
     
     model.set_env(test_env)
-    model.learn(total_timesteps = 100)
     assert(model.n_batch == 10)
-    os.remove(tmp_path / "a2c_cartpole.zip")
+    os.remove(os.path.join(tmp_path , "a2c_cartpole.zip"))

--- a/tests/test_a2c.py
+++ b/tests/test_a2c.py
@@ -13,13 +13,13 @@ def test_a2c_update_n_batch_on_load(tmp_path):
     model = A2C('MlpPolicy', env, n_steps = 10)
     
     model.learn(total_timesteps = 100)
-    model.save(os.path.join(tmp_path , "a2c_cartpole.zip"))
+    model.save(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))
     
     del model
     
-    model = A2C.load(os.path.join(tmp_path , "a2c_cartpole.zip"))
+    model = A2C.load(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))
     test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
     
     model.set_env(test_env)
     assert(model.n_batch == 10)
-    os.remove(os.path.join(tmp_path , "a2c_cartpole.zip"))
+    os.remove(os.path.join(str(tmp_path) , "a2c_cartpole.zip"))

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -1,8 +1,11 @@
 import os
 
 import pytest
+import gym 
 
 from stable_baselines import PPO2
+from stable_baselines.common import make_vec_env
+from stable_baselines.common.vec_env import DummyVecEnv
 
 
 @pytest.mark.parametrize("cliprange", [0.2, lambda x: 0.1 * x])
@@ -19,3 +22,18 @@ def test_clipping(cliprange, cliprange_vf):
 
     if os.path.exists('./ppo2_clip.zip'):
         os.remove('./ppo2_clip.zip')
+
+def test_update_n_batch_on_load():
+    env = make_vec_env('CartPole-v1',n_envs=2)
+    model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
+    
+    model.learn(total_timesteps = 100)
+    model.save("ppo2_cartpole")
+    
+    del model
+    
+    model = PPO2.load("ppo2_cartpole")
+    test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
+    
+    model.set_env(test_env)
+    model.learn(total_timesteps = 100)

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -28,13 +28,13 @@ def test_ppo2_update_n_batch_on_load(tmp_path):
     model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
     
     model.learn(total_timesteps = 100)
-    model.save(os.path.join(tmp_path , "ppo2_cartpole.zip"))
+    model.save(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))
     
     del model
     
-    model = PPO2.load(os.path.join(tmp_path , "ppo2_cartpole.zip"))
+    model = PPO2.load(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))
     test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
     
     model.set_env(test_env)
     model.learn(total_timesteps = 100)
-    os.remove(os.path.join(tmp_path , "ppo2_cartpole.zip"))
+    os.remove(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -37,3 +37,5 @@ def test_update_n_batch_on_load():
     
     model.set_env(test_env)
     model.learn(total_timesteps = 100)
+    if os.path.exists('./ppo2_cartpole.zip'):
+        os.remove('./ppo2_cartpole.zip')

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -1,4 +1,5 @@
 import os
+import pathlib
 
 import pytest
 import gym 
@@ -23,19 +24,18 @@ def test_clipping(cliprange, cliprange_vf):
     if os.path.exists('./ppo2_clip.zip'):
         os.remove('./ppo2_clip.zip')
 
-def test_update_n_batch_on_load():
+def test_update_n_batch_on_load(tmp_path):
     env = make_vec_env('CartPole-v1',n_envs=2)
     model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
     
     model.learn(total_timesteps = 100)
-    model.save("ppo2_cartpole")
+    model.save(tmp_path / "ppo2_cartpole.zip")
     
     del model
     
-    model = PPO2.load("ppo2_cartpole")
+    model = PPO2.load(tmp_path / "ppo2_cartpole.zip")
     test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
     
     model.set_env(test_env)
     model.learn(total_timesteps = 100)
-    if os.path.exists('./ppo2_cartpole.zip'):
-        os.remove('./ppo2_cartpole.zip')
+    os.remove(tmp_path / "ppo2_cartpole.zip")

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -1,5 +1,4 @@
 import os
-import pathlib
 
 import pytest
 import gym 
@@ -29,13 +28,13 @@ def test_ppo2_update_n_batch_on_load(tmp_path):
     model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
     
     model.learn(total_timesteps = 100)
-    model.save(tmp_path / "ppo2_cartpole.zip")
+    model.save(os.path.join(tmp_path , "ppo2_cartpole.zip"))
     
     del model
     
-    model = PPO2.load(tmp_path / "ppo2_cartpole.zip")
+    model = PPO2.load(os.path.join(tmp_path , "ppo2_cartpole.zip"))
     test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
     
     model.set_env(test_env)
     model.learn(total_timesteps = 100)
-    os.remove(tmp_path / "ppo2_cartpole.zip")
+    os.remove(os.path.join(tmp_path , "ppo2_cartpole.zip"))

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -1,7 +1,7 @@
 import os
 
 import pytest
-import gym 
+import gym
 
 from stable_baselines import PPO2
 from stable_baselines.common import make_vec_env
@@ -10,31 +10,38 @@ from stable_baselines.common.vec_env import DummyVecEnv
 
 @pytest.mark.parametrize("cliprange", [0.2, lambda x: 0.1 * x])
 @pytest.mark.parametrize("cliprange_vf", [None, 0.2, lambda x: 0.3 * x, -1.0])
-def test_clipping(cliprange, cliprange_vf):
+def test_clipping(tmp_path, cliprange, cliprange_vf):
     """Test the different clipping (policy and vf)"""
-    model = PPO2('MlpPolicy', 'CartPole-v1',
-                 cliprange=cliprange, cliprange_vf=cliprange_vf,
-                 noptepochs=2, n_steps=64).learn(100)
-    model.save('./ppo2_clip.zip')
+    model = PPO2(
+        "MlpPolicy",
+        "CartPole-v1",
+        cliprange=cliprange,
+        cliprange_vf=cliprange_vf,
+        noptepochs=2,
+        n_steps=64,
+    ).learn(100)
+    save_path = os.path.join(str(tmp_path), "ppo2_clip.zip")
+    model.save(save_path)
     env = model.get_env()
-    model = PPO2.load('./ppo2_clip.zip', env=env)
+    model = PPO2.load(save_path, env=env)
     model.learn(100)
 
-    if os.path.exists('./ppo2_clip.zip'):
-        os.remove('./ppo2_clip.zip')
+    if os.path.exists(save_path):
+        os.remove(save_path)
+
 
 def test_ppo2_update_n_batch_on_load(tmp_path):
-    env = make_vec_env('CartPole-v1',n_envs=2)
-    model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
-    
-    model.learn(total_timesteps = 100)
-    model.save(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))
-    
+    env = make_vec_env("CartPole-v1", n_envs=2)
+    model = PPO2("MlpPolicy", env, n_steps=10, nminibatches=1)
+    save_path = os.path.join(str(tmp_path), "ppo2_cartpole.zip")
+
+    model.learn(total_timesteps=100)
+    model.save(save_path)
+
     del model
-    
-    model = PPO2.load(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))
-    test_env = DummyVecEnv([lambda: gym.make('CartPole-v1')])
-    
+
+    model = PPO2.load(save_path)
+    test_env = DummyVecEnv([lambda: gym.make("CartPole-v1")])
+
     model.set_env(test_env)
-    model.learn(total_timesteps = 100)
-    os.remove(os.path.join(str(tmp_path) , "ppo2_cartpole.zip"))
+    model.learn(total_timesteps=100)

--- a/tests/test_ppo2.py
+++ b/tests/test_ppo2.py
@@ -24,7 +24,7 @@ def test_clipping(cliprange, cliprange_vf):
     if os.path.exists('./ppo2_clip.zip'):
         os.remove('./ppo2_clip.zip')
 
-def test_update_n_batch_on_load(tmp_path):
+def test_ppo2_update_n_batch_on_load(tmp_path):
     env = make_vec_env('CartPole-v1',n_envs=2)
     model = PPO2('MlpPolicy', env, n_steps = 10, nminibatches=1)
     

